shakedown/dcos/__init__.py

```diff
@@ -1,9 +1,107 @@
 import os
-import dcos
-import dcos.cluster
 import sys
 
+import select
+import threading
+
+import dcos
+import dcos.cluster
 import shakedown
+from .helpers import validate_key, try_close, get_transport, start_transport
+
+
+class TransportManager(object):
+    
+    def __init__(self):
+        self.connections = {}
+        self.mutex = threading.RLock()
+    
+    @staticmethod
+    def key_name(host, username):
+        return "{h}-{u}".format(h=host, u=username)
+    
+    @staticmethod
+    def _check_username(username):
+        return shakedown.cli.ssh_user if not username else username
+
+    def _open_transport(self, host, username, key_path):
+        """Open a new SSH transport/connection to host. This operation
+        is heavy, as it includes authenticating.
+
+        :param host: host or IP of the machine
+        :type host: str
+        :param username: SSH username
+        :type username: str
+        :param key_path: path to the SSH private key for SSH auth
+        :type key_path: str
+        :param noisy: verbose output
+        :type noisy: bool
+        :return: new, connected transport or None
+        """
+        username = self._check_username(username)
+        if not key_path:
+            key_path = shakedown.cli.ssh_key_file
+        
+        key = validate_key(key_path)
+        transport = get_transport(host, username, key)
+    
+        if transport:
+            transport = start_transport(transport, username, key)
+        else:
+            print("error: unable to connect to {}".format(host))
+            return None
+        
+        if transport.is_authenticated():
+            needle = self.key_name(host, username)
+            with self.mutex:
+                # make extra sure nothing else added this needle.
+                # if it did, use the existing one and close this one.
+                close_this_one = needle in self.connections
+                if not close_this_one:
+                    self.connections[needle] = transport
+                returnable = self.connections[needle]
+            
+            if close_this_one:
```

Comment:
if the connection is already open by another.. why would you close it?   It would be the responsibility of the owner to close it and wouldn't you loop back and get another?

Reply:
this logic in general is odd... 
you are getting a `returnable` even if `close_this_one` is true... I don't understand.

Reply:
Short version: to prevent unmanaged / dangling connections

This opens a new connection and before updating `self.connections` with the new connection, it double checks to make sure one still does not exist. If one did exist and new connection overwrote it, then the connection that existed would hang around and only be closed when shakedown exited.

This checks to make sure nothing was added and if one was, this closes the new connection (`transport`) it just opened and returns the connection it just discovered.